### PR TITLE
Fix sending event messages to non-subscribers

### PR DESCRIPTION
**Pull request by lwatt**
Closes https://github.com/dask/distributed/issues/7013

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @lwatt! Could you add a test to ensure that the changes here fix the original issue reported in https://github.com/dask/distributed/issues/7013?

**Comment by lwatt** [comment]
Thanks - I've added a test now. Confirmed this fails on the original version, and passes now.

**Comment by lwatt** [comment]
I've just pushed the linting updates from `pre-commit`. It would be great if this could be merged soon

File: distributed/tests/test_client.py [hunk 1]
```diff
@@ -6600,6 +6600,33 @@ def log_scheduler(dask_scheduler):
...
+    while len(s.event_subscriber["test-topic"]) != 2:
+        await asyncio.sleep(0.01)
+
+    await c.log_event("test-topic", {})
+
+    while len(received_events) < 2:
+        await asyncio.sleep(0.01)
+
+    assert len(received_events) == 2
+    assert {ev[0] for ev in received_events} == {1, 2}
```
Comment (hendrikmakait) [inline distributed/tests/test_client.py:6627]: From what I understand from #7013, having a non-subscribed client around caused a `ValueError before. Let's add it to the test case then:

```suggestion
        async with Client(s.address, asynchronous=True) as c2, Client(
            s.address, asynchronous=True
        ) as c3:
            received_events = []

            def get_event_handler(handler_id):
                def handler(event):
                    received_events.append((handler_id, event))

                return handler

            c.subscribe_topic("test-topic", get_event_handler(1))
            c2.subscribe_topic("test-topic", get_event_handler(2))

            while len(s.event_subscriber["test-topic"]) != 2:
                await asyncio.sleep(0.01)

            await c.log_event("test-topic", {})

            import time

            time.sleep(2)
            while len(received_events) < 2:
                await asyncio.sleep(0.01)

            assert len(received_events) == 2
            assert {handler_id for handler_id, _ in received_events} == {1, 2}
            assert "ValueError" not in logger.getvalue()
```
In this suggestion, I have made the following adjustments:
1. Add the non-subscribed client
2. Use an async context manager to ensure teardown of the clients
3. Capture the client logs to assert that the `ValueError` from #7013 is not logged
4. Stylistic nit to the set comparison

**Review by hendrikmakait — COMMENTED** [review]
Thanks, @lwatt for adding this fix. This change generally looks good to me, I have one suggestion for improving the test case you added.

File: distributed/tests/test_client.py (hunk 1 above)
Comment (lwatt) [reply distributed/tests/test_client.py:6627]: Thanks - Looks good, but do we need this part? 

```
import time

time.sleep(2)
```

Comment (hendrikmakait) [reply distributed/tests/test_client.py:6627]: Whoops, good catch, that snuck in during copy+paste. I've adjusted the suggestion.

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -6600,6 +6600,35 @@ def log_scheduler(dask_scheduler):
     assert events[1][1] == ("alice", "bob")
 
 
+@gen_cluster(client=True, nthreads=[])
```
Comment (hendrikmakait) [inline distributed/tests/test_client.py:6603]: It looks like a bunch of `+` at the beginnings of the lines got in here when applying the suggestion which is causing CI to fail.

Comment (lwatt) [reply distributed/tests/test_client.py:6603]: It was an indentation issue, I've fixed it

**Review by hendrikmakait — APPROVED** [review]
LGTM, thanks @lwatt, and sorry for the hiccup with the suggestion. CI failures appear to be unrelated.

**Comment by hiiwave** [comment]
@lwatt I was bothered by this today. Thanks for the fix!
I see this PR was approved. Could you proceed to merge it?

**Pull request merged**